x-pack/plugins/lens/public/indexpattern_datasource/operations/layer_helpers.ts

```diff
@@ -30,6 +30,81 @@ interface ColumnChange {
   shouldResetLabel?: boolean;
 }
 
+interface ColumnCopy {
+  layer: IndexPatternLayer;
+  columnId: string;
+  sourceColumn: IndexPatternColumn;
+  sourceColumnId: string;
+  shouldDeleteSource?: boolean;
+  indexPattern: IndexPattern;
+}
+
+export function copyColumn({
+  layer,
+  columnId,
+  sourceColumn,
+  sourceColumnId,
+  shouldDeleteSource,
+  indexPattern,
+}: ColumnCopy): IndexPatternLayer {
+  let newColumns = {
+    ...layer.columns,
+    [columnId]: { ...sourceColumn },
+  };
+
+  const operationDefinition = operationDefinitionMap[sourceColumn.operationType];
+  const newIds: string[] = [];
+  if (operationDefinition.input === 'fullReference' && 'references' in sourceColumn) {
+    sourceColumn?.references.forEach((ref) => {
+      const newId = generateId();
+      newIds.push(newId);
+      const refId = Object.keys(layer.columns).find((c) => c === ref);
+      if (refId) {
+        newColumns = {
+          ...newColumns,
+          [newId]: { ...layer.columns[refId] },
+          [columnId]: { ...sourceColumn, references: [newId] },
+        };
+      }
+    });
+  }
+
+  const newColumnOrder = [...layer.columnOrder];
+
+  if (shouldDeleteSource) {
+    const sourceIndex = newColumnOrder.findIndex((c) => c === sourceColumnId);
+    const targetIndex = newColumnOrder.findIndex((c) => c === columnId);
+
+    if (targetIndex === -1) {
+      // for newly created columns, remove the old entry and add the last one to the end
+      newColumnOrder.splice(sourceIndex, 1);
+      if (newIds.length) {
+        newColumnOrder.push(...newIds);
+      }
+      newColumnOrder.push(columnId);
```

Comment:
I think you should be calling `getColumnOrder` to derive the order from the columns, which will automatically add these new IDs into the correct grouping. The main reason is that adding metrics to the end off the list breaks our current assumptions that references are topologically sorted to be _after_ the thing they depend on